reasoner/resolution/resolver/SubsumptionTracker.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (C) 2021 Vaticle
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package com.vaticle.typedb.core.reasoner.resolution.resolver;
+
+import com.vaticle.typedb.core.concept.Concept;
+import com.vaticle.typedb.core.concept.answer.ConceptMap;
+import com.vaticle.typedb.core.traversal.common.Identifier;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+class SubsumptionTracker {
+
+    private final Map<ConceptMap, Set<ConceptMap>> subsumersMap;
+    private final Set<ConceptMap> finishedStates;
+    private final Map<ConceptMap, ConceptMap> finishedMapping;
+
+    public SubsumptionTracker() {
+        this.finishedStates = new HashSet<>();
+        this.subsumersMap = new HashMap<>();
+        this.finishedMapping = new HashMap<>();
+    }
+
+    public void addFinished(ConceptMap conceptMap) {
+        this.finishedStates.add(conceptMap);
+    }
+
+    public Optional<ConceptMap> getSubsumer(ConceptMap unfinished) {
```

Comment:
`getSubsumer` doesn't sound like it has any notion of finish/unfinished. 
This class from outside appears to have only to do with concept maps being subsets of other ones, right?

Maybe you already want to turn this into a `SubsumptiveCoordinator` which possibly has a `SubsumptionTracker` in it? The `Tracker` only knows about relationships between concept maps. 
The coordinator keeps track of finished/unfinished states and is roughly aware of "resolvers" answering queries and being complete or incomplete.

In other words the story/abstraction still feels a bit rough. How would you explain it in words?

Reply:
The `SubsumptionTracker` class isn't doing much if it isn't tracking which subsumers are finished and which finished subsumers we've already found in the past. I've moved it to be an inner class and tweaked `getSubsumer` to `getFinishedSubsumer` for clarity. Unsure of the next move.